packages/codecs-core/src/reverse-codec.ts

```diff
@@ -36,13 +59,16 @@ export function reverseDecoder<TTo, TSize extends number>(
     return createDecoder({
         ...decoder,
         read: (bytes, offset) => {
-            const reverseEnd = offset + decoder.fixedSize;
-            if (offset === 0 && bytes.length === reverseEnd) {
-                return decoder.read(new Uint8Array([...bytes]).reverse(), offset);
-            }
-            const reversedBytes = bytes.slice();
-            reversedBytes.set(bytes.slice(offset, reverseEnd).reverse(), offset);
-            return decoder.read(reversedBytes, offset);
+            const reversedBytes = new Uint8Array(decoder.fixedSize);
+            copySourceToTargetInReverse(
+                bytes /* source */,
+                reversedBytes /* target_WILL_MUTATE */,
+                offset /* sourceOffset */,
+                offset + decoder.fixedSize /* sourceLength */,
+                -offset /* targetOffset */,
+            );
+            const [decodedBytes, decodedOffset] = decoder.read(reversedBytes, 0);
+            return [decodedBytes, offset + decodedOffset];
```

Comment:
There is a slight issue here which is not obvious to see and tricky to test.

The issue is that the encoding logic is not completely symmetrical with the decoding logic.

Before, both the encoding and decoding were reversing bytes in situ. Meaning both the inner `encoder` and `decoder` would have access to the entirety of the byte array before/after the bytes have been reversed respectively.

After this PR, the encoding logic stays the same — i.e. the inner `encoder` has access to the full byte array — but **the decoding logic now creates a "box"** that only give access to the reversed bytes to the inner `decoder`.

This means that, if we had a `offsetCodec` inside of that inner codec, the results would be different when encoding and decoding.

![CleanShot 2024-05-27 at 09 24 48@2x](https://github.com/solana-labs/solana-web3.js/assets/3642397/0c28375d-8c9b-4a12-aae7-36393db3e17c)

Whatever we do, we need to ensure encoding and decoding with an `offsetCodec` inside yields the same result.

Here we can either box both functions or update the decoding logic so the inner `decoder` has access to the whole byte array such that part of it has been reversed. I am in favour of doing the latter since I think we should avoid creating boxes unless we really need to — e.g. `fixCodecSize`, `addCodecSizePrefix` and `addCodecSentinel` all need to be boxed.

We should probably add a test as well to ensure the codec is symmetrical with an `offsetCodec` inside the inner codec.

Reply:
Amazing review. Thank you!

Fixed.

Also tested.

Reply:
Also updated the benchmarks.